protos/core.proto

```diff
@@ -260,24 +267,414 @@ message CompiledGraph {
+    // Description of the assertion.
+    string description = 9;
+  }
+
+  message Operation {
+    // Target is the unique action identifier.
+    Target target = 1;
+
+    // Canonical Target is the unique action identifier, with no target
+    // overrides applied.
+    // This option cannot be manually specified in action configs.
+    Target canonical_target = 2;
+
+    // Targets (non-canonical) of actions that this action is dependent on.
+    repeated Target dependency_targets = 3;
+
+    // Path to the source file that the contents of the action is loaded from.
+    string source = 4;
+
+    // Path that the config of the action is loaded from.
+    // This option cannot be manually specified in action configs.
+    string config_path = 5;
+
+    // A list of user-defined tags with which the action should be labeled.
+    repeated string tags = 6;
+
+    // If set to true, this action will not be executed. However, the action may
+    // still be depended upon.
+    bool disabled = 7;
+
+    // The SQL query.
+    string query = 8;
+
+    // Declares that this action creates a dataset which should be referenceable
+    // as a dependency target, for example by using the `ref` function.
+    bool has_output = 9;
+
+    // Description of the operation.
+    string description = 10;
+
+    // Descriptions of columns within the operation. Can only be set if
+    // hasOutput is true.
+    repeated ColumnDescriptor columns = 11;
+  }
+
+  message Declaration {
+    // Target is the unique action identifier.
+    Target target = 1;
+
+    // Canonical Target is the unique action identifier, with no target
+    // overrides applied.
+    // This option cannot be manually specified in action configs.
+    Target canonical_target = 2;
+
+    // Targets (non-canonical) of actions that this action is dependent on.
+    repeated Target dependency_targets = 3;
+
+    // Path to the source file that the contents of the action is loaded from.
+    string source = 4;
+
+    // Path that the config of the action is loaded from.
+    // This option cannot be manually specified in action configs.
+    string config_path = 5;
+
+    // A list of user-defined tags with which the action should be labeled.
+    repeated string tags = 6;
+
+    // Description of the declaration.
+    string description = 7;
+  }
+
+  message Notebook {
+    // Target is the unique action identifier.
+    Target target = 1;
+
+    // Canonical Target is the unique action identifier, with no target
+    // overrides applied.
+    // This option cannot be manually specified in action configs.
+    Target canonical_target = 2;
+
+    // Targets (non-canonical) of actions that this action is dependent on.
+    repeated Target dependency_targets = 3;
+
+    // Path to the source file that the contents of the action is loaded from.
+    string source = 4;
+
+    // Path that the config of the action is loaded from.
+    // This option cannot be manually specified in action configs.
```

Comment:
I'm confused by these comments. is it because you intend these to be used in `actions.yaml` too?

if so... i know you won't like this, but I think this is confusing. i know you're trying to use the same protos in as many places as possible... but i think the cure is worse than the disease (of having multiple definitions).

i think it's confusing to have some fields (e.g. `config_path`) that cannot be set, but which exist in the proto. IMO it should be possible to generate help docs for `actions.yaml` schema without having to specify that some fields may not be set. (and be forced to check that those fields are not set in our code, etc)

Reply:
Yes, I think this makes a lot more sense.

I don't think it's that confusing, it's like putting an `[output_only = true]` option - which we do in the cloud API (there's just no proto field for this supported by default).

Having separate configs means:
* Users will have to look at the behavior of both the config option, and the outputted field - which in most cases will be identical. It's a LOT of information for a new user already, this would make it worse rather than better that the current situation.
* Hundreds more lines of proto definitions, requiring identical comments of the fields.
* Hundreds more lines of copying these fields from the config definitions to the action definitions.

Reply:
Moving to separate config protos, from discussion